libcudacxx/include/cuda/std/__numeric/mul_sat.h

```diff
@@ -0,0 +1,382 @@
+        {
+          return _CUDA_VSTD::numeric_limits<_Tp>::max();
+        }
+      }
+      else if (__x < _Tp{0} && __y > _Tp{0})
+      {
+        if (__x < static_cast<_Tp>(_CUDA_VSTD::numeric_limits<_Tp>::min() / __y))
+        {
+          return _CUDA_VSTD::numeric_limits<_Tp>::min();
+        }
+      }
+      else
+      {
+        if (__x > static_cast<_Tp>(_CUDA_VSTD::numeric_limits<_Tp>::min() / __y))
+        {
+          return _CUDA_VSTD::numeric_limits<_Tp>::min();
+        }
+      }
+    }
+    else
+    {
+      if (__x > static_cast<_Tp>(_CUDA_VSTD::numeric_limits<_Tp>::max() / __y))
+      {
+        return _CUDA_VSTD::numeric_limits<_Tp>::max();
+      }
+    }
+    return static_cast<_Tp>(__x * __y);
+  }
+
+#if defined(_CCCL_BUILTIN_MUL_OVERFLOW)
+  template <class _Tp>
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Tp __impl_builtin(_Tp __x, _Tp __y) noexcept
+  {
+    _Tp __result{};
+    bool __overflow = _CCCL_BUILTIN_MUL_OVERFLOW(__x, __y, &__result);
+
+    return __clamp_overflow<_Tp>(__x, __y, __result, __overflow);
+  }
+#endif // _CCCL_BUILTIN_MUL_OVERFLOW
+#if !_CCCL_COMPILER(NVRTC)
+  template <class _Tp>
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI _CCCL_HOST static _Tp __impl_host(_Tp __x, _Tp __y) noexcept
+  {
+    if constexpr (_CCCL_TRAIT(is_signed, _Tp))
+    {
+      if ((__x == _Tp{-1} && __y == _CUDA_VSTD::numeric_limits<_Tp>::min())
+          || (__y == _Tp{-1} && __x == _CUDA_VSTD::numeric_limits<_Tp>::min()))
+      {
+        return _CUDA_VSTD::numeric_limits<_Tp>::max();
+      }
+
+      if constexpr (sizeof(_Tp) == 1)
+      {
+#  if _CCCL_COMPILER(MSVC, >=, 19, 37) && _CCCL_ARCH(X86_64)
+        int16_t __result;
+        bool __overflow = _mul_full_overflow_i8(__x, __y, &__result);
+        return __clamp_overflow<int8_t>(__x, __y, static_cast<int8_t>(__result), __overflow);
+#  else
+        return __impl_generic(__x, __y);
+#  endif
+      }
+      else if constexpr (sizeof(_Tp) == 2)
+      {
+#  if _CCCL_COMPILER(MSVC, >=, 19, 37) && _CCCL_ARCH(X86_64)
+        int16_t __result;
+        bool __overflow = _mul_overflow_i16(__x, __y, &__result);
+        return __clamp_overflow<int16_t>(__x, __y, __result, __overflow);
+#  else
+        return __impl_generic(__x, __y);
+#  endif
+      }
+      else if constexpr (sizeof(_Tp) == 4)
+      {
+#  if _CCCL_COMPILER(MSVC, >=, 19, 37) && _CCCL_ARCH(X86_64)
+        int32_t __result;
+        bool __overflow = _mul_overflow_i32(__x, __y, &__result);
+        return __clamp_overflow<int32_t>(__x, __y, __result, __overflow);
+#  elif _CCCL_COMPILER(MSVC) && _CCCL_ARCH(X86_64)
+        const int64_t __result = __emul(__x, __y);
+        return __clamp_overflow<int32_t>(
+          __x,
+          __y,
+          static_cast<int32_t>(__result),
+          __result > _CUDA_VSTD::numeric_limits<int32_t>::max()
+            || __result < _CUDA_VSTD::numeric_limits<int32_t>::min());
+#  else
```

Comment:
I would like to have comments here and other functions that show what branch this is. Its a bit difficult to see on a glance that we are in the !MSVC branch